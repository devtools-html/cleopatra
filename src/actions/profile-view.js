/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type {
  ProfileSelection, CallTreeFilter, ImplementationFilter,
} from '../types/actions';
import type { Action, ThunkAction } from '../types/store';
import type { ThreadIndex, IndexIntoFuncTable, IndexIntoMarkersTable } from '../types/profile';

/**
 * The actions that pertain to changing the view on the profile, including searching
 * and filtering. Currently the call tree's actions are in this file, but should be
 * split apart. These actions should most likely affect every panel.
 */
export function changeSelectedFuncStack(
  threadIndex: ThreadIndex,
  selectedFuncStack: IndexIntoFuncTable[]
): Action {
  return {
    type: 'CHANGE_SELECTED_FUNC_STACK',
    selectedFuncStack, threadIndex,
  };
}

export function changeSelectedThread(selectedThread: ThreadIndex): Action {
  return {
    type: 'CHANGE_SELECTED_THREAD',
    selectedThread,
  };
}

export function changeThreadOrder(threadOrder: ThreadIndex[]): Action {
  return {
    type: 'CHANGE_THREAD_ORDER',
    threadOrder,
  };
}

export function hideThread(
  threadIndex: ThreadIndex,
  threadOrder: ThreadIndex[],
  hiddenThreads: ThreadIndex[]
): ThunkAction<void> {
  return dispatch => {
    // Do not allow hiding the last thread.
    if (hiddenThreads.length + 1 === threadOrder.length) {
      return;
    }

    dispatch(({
      type: 'HIDE_THREAD',
      threadIndex,
      threadOrder,
      hiddenThreads,
    }: Action));
  };
}

export function showThread(threadIndex: ThreadIndex): Action {
  return {
    type: 'SHOW_THREAD',
    threadIndex,
  };
}

export function changeCallTreeSearchString(searchString: string): Action {
  return {
    type: 'CHANGE_CALL_TREE_SEARCH_STRING',
    searchString,
  };
}

export function changeExpandedFuncStacks(
  threadIndex: ThreadIndex,
  expandedFuncStacks: Array<IndexIntoFuncTable[]>
): Action {
  return {
    type: 'CHANGE_EXPANDED_FUNC_STACKS',
    threadIndex, expandedFuncStacks,
  };
}

export function changeSelectedMarker(
  threadIndex: ThreadIndex,
  selectedMarker: IndexIntoMarkersTable | -1
): Action {
  return {
    type: 'CHANGE_SELECTED_MARKER',
    selectedMarker, threadIndex,
  };
}

export function changeImplementationFilter(implementation: ImplementationFilter): Action {
  return {
    type: 'CHANGE_IMPLEMENTATION_FILTER',
    implementation,
  };
}

export function changeInvertCallstack(invertCallstack: boolean): Action {
  return {
    type: 'CHANGE_INVERT_CALLSTACK',
    invertCallstack,
  };
}

export function changeHidePlatformDetails(hidePlatformDetails: boolean): Action {
  return {
    type: 'CHANGE_HIDE_PLATFORM_DETAILS',
    hidePlatformDetails,
  };
}

export type UpdateProfileSelection = (selection: ProfileSelection) => Action;

export function updateProfileSelection(selection: ProfileSelection): Action {
  return {
    type: 'UPDATE_PROFILE_SELECTION',
    selection,
  };
}

export function addRangeFilter(start: number, end: number): Action {
  return {
    type: 'ADD_RANGE_FILTER',
    start, end,
  };
}

export function addRangeFilterAndUnsetSelection(start: number, end: number): ThunkAction<void> {
  return dispatch => {
    dispatch(addRangeFilter(start, end));
    dispatch(updateProfileSelection({ hasSelection: false, isModifying: false }));
  };
}

export function popRangeFilters(firstRemovedFilterIndex: number): Action {
  return {
    type: 'POP_RANGE_FILTERS',
    firstRemovedFilterIndex,
  };
}

export function popRangeFiltersAndUnsetSelection(firstRemovedFilterIndex: number): ThunkAction<void> {
  return dispatch => {
    dispatch(popRangeFilters(firstRemovedFilterIndex));
    dispatch(updateProfileSelection({ hasSelection: false, isModifying: false }));
  };
}

export function addCallTreeFilter(threadIndex: ThreadIndex, filter: CallTreeFilter): Action {
  return {
    type: 'ADD_CALL_TREE_FILTER',
    threadIndex,
    filter,
  };
}

export function popCallTreeFilters(threadIndex: ThreadIndex, firstRemovedFilterIndex: number): Action {
  return {
    type: 'POP_CALL_TREE_FILTERS',
    threadIndex,
    firstRemovedFilterIndex,
  };
}
