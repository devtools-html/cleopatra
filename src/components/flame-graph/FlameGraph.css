/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.flameGraph {
  flex: 1;
  display: flex;
  flex-flow: column;
  overflow: hidden;
}

.flameGraphContent {
  display: flex;
  flex-direction: row;
  flex: 1;
  border-top: 1px solid var(--grey-30);
}
