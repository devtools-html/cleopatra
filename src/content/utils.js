// @flow
type ValidWeakMapKey = Object | [];

/**
 * Memoize a function's results using a WeakMap and its first parameter.
 */
export function weakMemoize(fn: (ValidWeakMapKey, ...any) => any) {
  const memo = new WeakMap();
  return function (key: ValidWeakMapKey) {
    const cachedValue = memo.get(key);
    if (cachedValue === undefined) {
      const value = fn.apply(null, arguments);
      memo.set(key, value);
      return value;
    }
    return cachedValue;
  };
}
