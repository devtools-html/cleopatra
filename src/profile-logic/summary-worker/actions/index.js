/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { summarizeProfile } from '../../../profile-logic/summarize-profile';

export function processProfileSummary() {
  return function (dispatch, getState) {
    dispatch({
      toContent: true,
      type: 'PROFILE_SUMMARY_PROCESSED',
      summary: summarizeProfile(getState().profile),
    });
  };
}

export function profileProcessed(profile) {
  return {
    type: 'PROFILE_PROCESSED',
    profile: profile,
  };
}
