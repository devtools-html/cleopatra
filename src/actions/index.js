/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import * as icons from './icons';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';
import * as profileUpload from './profile-upload';
import * as shortUrl from './short-url';
import * as urlState from './url-state';

export default Object.assign(
  {},
  app,
  icons,
  profileSummary,
  profileUpload,
  profileView,
  receiveProfile,
  shortUrl,
  timeline,
  urlState
);
