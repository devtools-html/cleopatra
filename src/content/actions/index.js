// @flow
import * as app from './app';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';

export default Object.assign({}, app, profileSummary, profileView, receiveProfile, timeline);
