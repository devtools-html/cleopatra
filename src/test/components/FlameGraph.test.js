/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import FlameGraph from '../../components/flame-graph';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import mockCanvasContext from '../fixtures/mocks/canvas-context';
import { storeWithProfile } from '../fixtures/stores';
import { getBoundingBox } from '../fixtures/utils';
import { getProfileFromTextSamples } from '../fixtures/profiles/make-profile';
import { changeInvertCallstack } from '../../actions/profile-view';
import mockRaf from '../fixtures/mocks/request-animation-frame';

it('renders FlameGraph correctly', () => {
  const flushRafCalls = mockRaf();
  window.devicePixelRatio = 1;
  const ctx = mockCanvasContext();

  /**
   * Mock out any created refs for the components with relevant information.
   */
  function createNodeMock(element) {
    // <FlameGraphCanvas><canvas /></FlameGraphCanvas>
    if (element.type === 'canvas') {
      return {
        getBoundingClientRect: () => getBoundingBox(200, 300),
        getContext: () => ctx,
        style: {},
      };
    }
    // <Viewport />
    if (element.props.className.split(' ').includes('chartViewport')) {
      return {
        getBoundingClientRect: () => getBoundingBox(200, 300),
        focus: () => {},
      };
    }
    return null;
  }

  const { profile } = getProfileFromTextSamples(`
    A  A  A
    B  B  B
    C  C  H
    D  F  I
    E  G
  `);

  const store = storeWithProfile(profile);

  const flameGraph = renderer.create(
    <Provider store={store}>
      <FlameGraph />
    </Provider>,
    { createNodeMock }
  );

  flushRafCalls();

  const drawCalls = ctx.__flushDrawLog();

  expect(flameGraph).toMatchSnapshot();
  expect(drawCalls).toMatchSnapshot();

  delete window.devicePixelRatio;
});

it('renders a message instead of FlameGraph when call stack is inverted', () => {
  const { profile } = getProfileFromTextSamples(`
    A  B
  `);

  const store = storeWithProfile(profile);
  store.dispatch(changeInvertCallstack(true));

  const flameGraph = renderer.create(
    <Provider store={store}>
      <FlameGraph />
    </Provider>
  );

  expect(flameGraph).toMatchSnapshot();
});
