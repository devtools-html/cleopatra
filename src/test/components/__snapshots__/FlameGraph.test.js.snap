// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders FlameGraph correctly 1`] = `
<div
  className="flameGraph"
>
  <div
    className="stackSettings"
  >
    <ul
      className="stackSettingsList"
    >
      <li
        className="stackSettingsListItem stackSettingsFilter"
      >
        <label
          className="stackSettingsFilterLabel"
        >
          <input
            checked={true}
            className="stackSettingsFilterInput"
            name="stack-settings-filter"
            onChange={[Function]}
            title="Filter stack frames to a type."
            type="radio"
            value="combined"
          />
          All stacks
        </label>
        <label
          className="stackSettingsFilterLabel"
        >
          <input
            checked={false}
            className="stackSettingsFilterInput"
            name="stack-settings-filter"
            onChange={[Function]}
            title="Filter stack frames to a type."
            type="radio"
            value="js"
          />
          JavaScript
        </label>
        <label
          className="stackSettingsFilterLabel"
        >
          <input
            checked={false}
            className="stackSettingsFilterInput"
            name="stack-settings-filter"
            onChange={[Function]}
            title="Filter stack frames to a type."
            type="radio"
            value="cpp"
          />
          Native
        </label>
      </li>
    </ul>
    <div
      className="stackSearchField stackSettingsSearchField"
    >
      <label
        className="stackSearchFieldLabel"
      >
        Filter stacks: 
        <form
          className="idleSearchField stackSearchFieldInput"
          onSubmit={[Function]}
        >
          <input
            className="idleSearchFieldInput"
            name="search"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required="required"
            title="Only display stacks which contain a function whose name matches this substring"
            type="search"
            value=""
          />
          <input
            className="idleSearchFieldButton"
            onClick={[Function]}
            onFocus={[Function]}
            type="reset"
          />
        </form>
        <div
          className="stackSearchFieldIntroduction isHidden"
        >
          Did you know you can use the comma (,) to search using several terms?
        </div>
      </label>
    </div>
  </div>
  <ol
    className="filterNavigatorBar calltreeTransformNavigator"
  >
    <li
      className="filterNavigatorBarItem filterNavigatorBarRootItem filterNavigatorBarSelectedItem filterNavigatorBarLeafItem"
      data-index={0}
      onClick={[Function]}
      title="Complete \\"Empty\\""
    >
      <span
        className="filterNavigatorBarItemContent"
      >
        Complete "Empty"
      </span>
    </li>
  </ol>
  <div
    className="flameGraphContent"
  >
    <div
      className="react-contextmenu-wrapper treeViewContextMenu"
      onContextMenu={[Function]}
      onMouseDown={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
    >
      <div
        className="chartViewport"
        onMouseDown={[Function]}
        onWheel={[Function]}
      >
        <div>
          <canvas
            className="chartCanvas flameGraphCanvas"
            onDoubleClick={[Function]}
            onMouseDown={[Function]}
            onMouseMove={[Function]}
            onMouseOut={[Function]}
            onMouseUp={[Function]}
          />
        </div>
        <div
          className="chartViewportShiftScroll hidden"
        >
          Zoom Chart:
          <kbd
            className="chartViewportShiftScrollKbd"
          >
            Shift
          </kbd>
          <kbd
            className="chartViewportShiftScrollKbd"
          >
            Scroll
          </kbd>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders FlameGraph correctly 2`] = `
Array [
  Array [
    "scale",
    1,
    1,
  ],
  Array [
    "measureText",
    "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ.()< /:-_",
  ],
  Array [
    "measureText",
    "â€¦",
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    0,
    0,
    200,
    300,
  ],
  Array [
    "set fillStyle",
    "#c1e0ff",
  ],
  Array [
    "fillRect",
    0,
    284,
    200,
    15,
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    0,
    284,
    1,
    15,
  ],
  Array [
    "measureText",
    "A",
  ],
  Array [
    "set fillStyle",
    "#000f40",
  ],
  Array [
    "fillText",
    "A",
    3,
    295,
  ],
  Array [
    "set fillStyle",
    "#c1e0ff",
  ],
  Array [
    "fillRect",
    0,
    268,
    200,
    15,
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    0,
    268,
    1,
    15,
  ],
  Array [
    "measureText",
    "B",
  ],
  Array [
    "set fillStyle",
    "#000f40",
  ],
  Array [
    "fillText",
    "B",
    3,
    279,
  ],
  Array [
    "set fillStyle",
    "#c1e0ff",
  ],
  Array [
    "fillRect",
    0,
    252,
    133.33333333333331,
    15,
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    0,
    252,
    1,
    15,
  ],
  Array [
    "measureText",
    "C",
  ],
  Array [
    "set fillStyle",
    "#000f40",
  ],
  Array [
    "fillText",
    "C",
    3,
    263,
  ],
  Array [
    "set fillStyle",
    "#c1e0ff",
  ],
  Array [
    "fillRect",
    133.33333333333331,
    252,
    66.66666666666667,
    15,
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    133.33333333333331,
    252,
    1,
    15,
  ],
  Array [
    "measureText",
    "H",
  ],
  Array [
    "set fillStyle",
    "#000f40",
  ],
  Array [
    "fillText",
    "H",
    136.33333333333331,
    263,
  ],
  Array [
    "set fillStyle",
    "#c1e0ff",
  ],
  Array [
    "fillRect",
    0,
    236,
    66.66666666666666,
    15,
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    0,
    236,
    1,
    15,
  ],
  Array [
    "measureText",
    "D",
  ],
  Array [
    "set fillStyle",
    "#000f40",
  ],
  Array [
    "fillText",
    "D",
    3,
    247,
  ],
  Array [
    "set fillStyle",
    "#c1e0ff",
  ],
  Array [
    "fillRect",
    66.66666666666666,
    236,
    66.66666666666666,
    15,
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    66.66666666666666,
    236,
    1,
    15,
  ],
  Array [
    "measureText",
    "F",
  ],
  Array [
    "set fillStyle",
    "#000f40",
  ],
  Array [
    "fillText",
    "F",
    69.66666666666666,
    247,
  ],
  Array [
    "set fillStyle",
    "#218fff",
  ],
  Array [
    "fillRect",
    133.33333333333331,
    236,
    66.66666666666667,
    15,
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    133.33333333333331,
    236,
    1,
    15,
  ],
  Array [
    "measureText",
    "I",
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillText",
    "I",
    136.33333333333331,
    247,
  ],
  Array [
    "set fillStyle",
    "#218fff",
  ],
  Array [
    "fillRect",
    0,
    220,
    66.66666666666666,
    15,
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    0,
    220,
    1,
    15,
  ],
  Array [
    "measureText",
    "E",
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillText",
    "E",
    3,
    231,
  ],
  Array [
    "set fillStyle",
    "#218fff",
  ],
  Array [
    "fillRect",
    66.66666666666666,
    220,
    66.66666666666666,
    15,
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    66.66666666666666,
    220,
    1,
    15,
  ],
  Array [
    "measureText",
    "G",
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillText",
    "G",
    69.66666666666666,
    231,
  ],
]
`;

exports[`renders a message instead of FlameGraph when call stack is inverted 1`] = `
<div
  className="flameGraph"
>
  <div
    className="stackSettings"
  >
    <ul
      className="stackSettingsList"
    >
      <li
        className="stackSettingsListItem stackSettingsFilter"
      >
        <label
          className="stackSettingsFilterLabel"
        >
          <input
            checked={true}
            className="stackSettingsFilterInput"
            name="stack-settings-filter"
            onChange={[Function]}
            title="Filter stack frames to a type."
            type="radio"
            value="combined"
          />
          All stacks
        </label>
        <label
          className="stackSettingsFilterLabel"
        >
          <input
            checked={false}
            className="stackSettingsFilterInput"
            name="stack-settings-filter"
            onChange={[Function]}
            title="Filter stack frames to a type."
            type="radio"
            value="js"
          />
          JavaScript
        </label>
        <label
          className="stackSettingsFilterLabel"
        >
          <input
            checked={false}
            className="stackSettingsFilterInput"
            name="stack-settings-filter"
            onChange={[Function]}
            title="Filter stack frames to a type."
            type="radio"
            value="cpp"
          />
          Native
        </label>
      </li>
    </ul>
    <div
      className="stackSearchField stackSettingsSearchField"
    >
      <label
        className="stackSearchFieldLabel"
      >
        Filter stacks: 
        <form
          className="idleSearchField stackSearchFieldInput"
          onSubmit={[Function]}
        >
          <input
            className="idleSearchFieldInput"
            name="search"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required="required"
            title="Only display stacks which contain a function whose name matches this substring"
            type="search"
            value=""
          />
          <input
            className="idleSearchFieldButton"
            onClick={[Function]}
            onFocus={[Function]}
            type="reset"
          />
        </form>
        <div
          className="stackSearchFieldIntroduction isHidden"
        >
          Did you know you can use the comma (,) to search using several terms?
        </div>
      </label>
    </div>
  </div>
  <ol
    className="filterNavigatorBar calltreeTransformNavigator"
  >
    <li
      className="filterNavigatorBarItem filterNavigatorBarRootItem filterNavigatorBarSelectedItem filterNavigatorBarLeafItem"
      data-index={0}
      onClick={[Function]}
      title="Complete \\"Empty\\""
    >
      <span
        className="filterNavigatorBarItemContent"
      >
        Complete "Empty"
      </span>
    </li>
  </ol>
  <div
    className="flameGraphDisabledMessage"
  >
    <h3>
      The Flame Graph is not available for inverted call stacks
    </h3>
    <p>
      <button
        onClick={[Function]}
        type="button"
      >
        Switch to the normal call stack
      </button>
       
      to show the Flame Graph.
    </p>
  </div>
</div>
`;
